### PR TITLE
Engine - Wazuh dashboard integration

### PR DESCRIPTION
**Pull request by nbertoldo**
|Related issue|
|---|
|#17582|

## Description
This integration must be able to process different types of events generated on the [Wazuh Dashboard](https://documentation.wazuh.com/current/installation-guide/wazuh-dashboard/index.html#wazuh-dashboard).
Wazuh dashboard provides management and monitoring capabilities, giving users control over the Wazuh infrastructure. You can monitor your agents status and configuration, query and visualize your alert data and monitor manager rules and configuration.

#### Type: log
```
Jun 23 19:26:45 engine opensearch-dashboards[211576]: {"type":"log","@timestamp":"2023-06-23T19:26:45Z","tags":["info","http","server","OpenSearchDashboards"],"pid":211576,"message":"http server running at https://0.0.0.0:443"}
Jun 23 19:26:44 engine opensearch-dashboards[211576]: {"type":"log","@timestamp":"2023-06-23T19:26:44Z","tags":["info","savedobjects-service"],"pid":211576,"message":"Waiting until all OpenSearch nodes are compatible with OpenSearch Dashboards before starting saved objects migrations..."}
Jun 23 15:10:54 engine opensearch-dashboards[681]: {"type":"log","@timestamp":"2023-06-23T15:10:54Z","tags":["error","plugins","securityDashboards"],"pid":681,"message":"Failed authentication: Error: Authentication Exception"}
Jun 23 12:25:00 engine opensearch-dashboards[681]: {"type":"log","@timestamp":"2023-06-23T12:25:00Z","tags":["error","plugins","wazuh","cron-scheduler"],"pid":681,"message":"Error: Request failed with status code 400"}
Jun 26 04:20:00 engine opensearch-dashboards[681]: {"type":"log","@timestamp":"2023-06-26T04:20:00Z","tags":["error","plugins","wazuh","cron-scheduler"],"pid":681,"message":"Error: connect ECONNREFUSED 127.0.0.1:55000"}
```
#### Type: error
```
Jun 23 14:58:39 engine opensearch-dashboards[681]: {"type":"error","@timestamp":"2023-06-23T14:58:39Z","tags":["connection","client","error"],"pid":681,"level":"error","error":{"message":"139698220353408:error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1544:SSL alert number 46\n","name":"Error","stack":"Error: 139698220353408:error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1544:SSL alert number 46\n","code":"ERR_SSL_SSLV3_ALERT_CERTIFICATE_UNKNOWN"},"message":"139698220353408:error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert certificate unknown:../deps/openssl/openssl/ssl/record/rec_layer_s3.c:1544:SSL alert number 46\n"}
Jun 26 18:45:02 engine opensearch-dashboards[681]: {"type":"error","@timestamp":"2023-06-26T18:45:02Z","tags":[],"pid":681,"level":"error","error":{"message":"Internal Server Error","name":"Error","stack":"Error: Internal Server Error\n    at HapiResponseAdapter.toError (/usr/share/wazuh-dashboard/src/core/server/http/router/response_adapter.js:143:19)\n    at HapiResponseAdapter.toHapiResponse (/usr/share/wazuh-dashboard/src/core/server/http/router/response_adapter.js:97:19)\n    at HapiResponseAdapter.handle (/usr/share/wazuh-dashboard/src/core/server/http/router/response_adapter.js:92:17)\n    at Router.handle (/usr/share/wazuh-dashboard/src/core/server/http/router/router.js:164:34)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at handler (/usr/share/wazuh-dashboard/src/core/server/http/router/router.js:124:50)\n    at exports.Manager.execute (/usr/share/wazuh-dashboard/node_modules/@hapi/hapi/lib/toolkit.js:60:28)\n    at Object.internals.handler (/usr/share/wazuh-dashboard/node_modules/@hapi/hapi/lib/handler.js:46:20)\n    at exports.execute (/usr/share/wazuh-dashboard/node_modules/@hapi/hapi/lib/handler.js:31:20)\n    at Request._lifecycle (/usr/share/wazuh-dashboard/node_modules/@hapi/hapi/lib/request.js:371:32)\n    at Request._execute (/usr/share/wazuh-dashboard/node_modules/@hapi/hapi/lib/request.js:281:9)"},"url":"https://192.168.56.10/api/request","message":"Internal Server Error"}
```
#### Type: response
```
Jun 23 15:18:26 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:18:26Z","tags":[],"pid":681,"method":"post","statusCode":200,"req":{"url":"/auth/login","method":"post","headers":{"host":"192.168.56.10","connection":"keep-alive","content-length":"47","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","content-type":"application/json","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","osd-version":"2.6.0","sec-ch-ua-platform":"\"Linux\"","accept":"*/*","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh"},"res":{"statusCode":200,"responseTime":8,"contentLength":9},"message":"POST /auth/login 200 8ms - 9.0B"}
Jun 23 15:18:27 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:18:27Z","tags":[],"pid":681,"method":"post","statusCode":200,"req":{"url":"/api/core/capabilities","method":"post","headers":{"host":"192.168.56.10","connection":"keep-alive","content-length":"434","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","content-type":"application/json","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","osd-version":"2.6.0","sec-ch-ua-platform":"\"Linux\"","accept":"*/*","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/wazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/wazuh"},"res":{"statusCode":200,"responseTime":4,"contentLength":9},"message":"POST /api/core/capabilities 200 4ms - 9.0B"}
Jun 23 15:18:28 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:18:28Z","tags":[],"pid":681,"method":"post","statusCode":401,"req":{"url":"/api/request","method":"post","headers":{"host":"192.168.56.10","connection":"keep-alive","content-length":"73","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","accept":"application/json, text/plain, */*","content-type":"application/json","osd-xsrf":"kibana","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","sec-ch-ua-platform":"\"Linux\"","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/wazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/wazuh"},"res":{"statusCode":401,"responseTime":448,"contentLength":9},"message":"POST /api/request 401 448ms - 9.0B"}
Jun 23 15:14:32 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:14:31Z","tags":[],"pid":681,"method":"post","statusCode":401,"req":{"url":"/auth/login","method":"post","headers":{"host":"192.168.56.10","connection":"keep-alive","content-length":"44","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","content-type":"application/json","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","osd-version":"2.6.0","sec-ch-ua-platform":"\"Linux\"","accept":"*/*","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh"},"res":{"statusCode":401,"responseTime":245,"contentLength":9},"message":"POST /auth/login 401 245ms - 9.0B"}
Jun 23 15:31:24 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:31:24Z","tags":[],"pid":681,"method":"post","statusCode":200,"req":{"url":"/auth/logout","method":"post","headers":{"host":"192.168.56.10","connection":"keep-alive","content-length":"0","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","content-type":"application/json","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","osd-version":"2.6.0","sec-ch-ua-platform":"\"Linux\"","accept":"*/*","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/wazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/wazuh"},"res":{"statusCode":200,"responseTime":2,"contentLength":9},"message":"POST /auth/logout 200 2ms - 9.0B"}
Jun 23 15:18:26 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T15:18:26Z","tags":[],"pid":681,"method":"get","statusCode":200,"req":{"url":"/app/wazuh","method":"get","headers":{"host":"192.168.56.10","connection":"keep-alive","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","sec-ch-ua-mobile":"?0","sec-ch-ua-platform":"\"Linux\"","upgrade-insecure-requests":"1","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7","sec-fetch-site":"same-origin","sec-fetch-mode":"navigate","sec-fetch-user":"?1","sec-fetch-dest":"document","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh"},"res":{"statusCode":200,"responseTime":22,"contentLength":9},"message":"GET /app/wazuh 200 22ms - 9.0B"}
Jun 23 14:58:39 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-23T14:58:39Z","tags":[],"pid":681,"method":"get","statusCode":302,"req":{"url":"/app/wazuh","method":"get","headers":{"host":"192.168.56.10","connection":"keep-alive","cache-control":"max-age=0","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","sec-ch-ua-mobile":"?0","sec-ch-ua-platform":"\"Linux\"","upgrade-insecure-requests":"1","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7","sec-fetch-site":"same-origin","sec-fetch-mode":"navigate","sec-fetch-user":"?1","sec-fetch-dest":"document","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/login?nextUrl=%2Fapp%2Fwazuh"},"res":{"statusCode":302,"responseTime":35,"contentLength":9},"message":"GET /app/wazuh 302 35ms - 9.0B"}
Jun 26 18:39:55 engine opensearch-dashboards[681]: {"type":"response","@timestamp":"2023-06-26T18:39:55Z","tags":[],"pid":681,"method":"get","statusCode":200,"req":{"url":"/api/saved_objects/_find?type=index-pattern&fields=title&fields=fields&per_page=9999","method":"get","headers":{"host":"192.168.56.10","connection":"keep-alive","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","id":"default","pattern":"wazuh-alerts-*","accept":"application/json, text/plain, */*","osd-xsrf":"kibana","sec-ch-ua-platform":"\"Linux\"","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/wazuh","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/wazuh"},"res":{"statusCode":200,"responseTime":26,"contentLength":9},"message":"GET /api/saved_objects/_find?type=index-pattern&fields=title&fields=fields&per_page=9999 200 26ms - 9.0B"}
Jun 23 20:46:39 engine opensearch-dashboards[211576]: {"type":"response","@timestamp":"2023-06-23T20:46:39Z","tags":[],"pid":211576,"method":"delete","statusCode":200,"req":{"url":"/api/v1/configuration/internalusers/test_admin","method":"delete","headers":{"host":"192.168.56.10","connection":"keep-alive","sec-ch-ua":"\"Chromium\";v=\"110\", \"Not A(Brand\";v=\"24\", \"Google Chrome\";v=\"110\"","content-type":"application/json","sec-ch-ua-mobile":"?0","user-agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","osd-version":"2.6.0","sec-ch-ua-platform":"\"Linux\"","accept":"*/*","origin":"https://192.168.56.10","sec-fetch-site":"same-origin","sec-fetch-mode":"cors","sec-fetch-dest":"empty","referer":"https://192.168.56.10/app/security-dashboards-plugin","accept-encoding":"gzip, deflate, br","accept-language":"es-419,es;q=0.9"},"remoteAddress":"192.168.56.1","userAgent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36","referer":"https://192.168.56.10/app/security-dashboards-plugin"},"res":{"statusCode":200,"responseTime":25,"contentLength":9},"message":"DELETE /api/v1/configuration/internalusers/test_admin 200 25ms - 9.0B"}
```
### Alphanumeric Parser
In order to obtain the `process.name` field which in this case contains the special character `'-'`, it is necessary to modify the alphanumeric parser so that the user can extend the set of allowed characters by passing as parameter a string containing the character(s) to be added. For example, given the following event:
```json
{
    "inputField": "_123abc-DEF456"
}
```
And the following operation:
```yml
parse:
  logpar:
    - inputField: <~outputField/alphanumeric/-_>
```
The event will be modified as follows:
```json
{
    "inputField": "_123abc-DEF456",
    "~outputField": "_123abc-DEF456"
}
```

## Task List:
- [x] Obtain a list and a pre-analysis of the possible event types with the default configuration.
- [x] Create a decoder to extract the useful fields of each event type.
- [x] Analysis of fields that can be obtained but there is no clear destination in the ECS.
- [x] Analysis of fields that cannot be extracted (parse). 
- [x] Generate end to end events and check filbeat logs to rule out errors
- [x] Add test and expected for all event types

### CODE REVIEW

File: src/engine/ruleset/integrations/wazuh-dashboard/README.md
[the remark this answers is not in the document]
Comment (nbertoldo) [reply src/engine/ruleset/integrations/wazuh-dashboard/README.md:1]: I agree. Done

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/source/hlp/src/syntax.hpp [hunk 1]
```diff
@@ -350,17 +350,32 @@ inline Parser toEnd(const std::vector<std::string>& endTokens)
...
+                    return abs::makeFailure<ResultT>(input, {});
+                }
+            }
+            else
+            {
+                return abs::makeFailure<ResultT>(input, {});
+            }
+        }
         return abs::makeSuccess<ResultT>(input.substr(1));
     };
```
[the remark this answers is not in the document]
Comment (nbertoldo) [reply src/engine/source/hlp/src/syntax.hpp]: Done

**Review by LucioDonda — APPROVED** [review]
LGTM!

**Review by juliancnn — APPROVED** [review]
LGTM!

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Pull request merged**